hal/src/nRF52840/core_hal.c

```diff
@@ -464,11 +486,229 @@ void HAL_Core_Enter_Bootloader(bool persist) {
+        hfclk_resume = true;
+        nrf_drv_clock_hfclk_release();
+        while (nrf_drv_clock_hfclk_is_running()) {
+            ;
+        }
+    }
+
+    uint32_t exit_conditions = 0x00;
+
+    // Suspend all GPIOTE interrupts
+    HAL_Interrupts_Suspend();
+
+    for (int i = 0; i < pins_count; i++) {
+        pin_t wakeUpPin = pins[i];
+        InterruptMode edgeTriggerMode = (i < mode_count) ? mode[i] : mode[mode_count - 1];
+
+        PinMode wakeUpPinMode = INPUT;
+        // Set required pinMode based on edgeTriggerMode
+        switch(edgeTriggerMode) {
+            case RISING: {
+                wakeUpPinMode = INPUT_PULLDOWN;
+                break;
+            }
+            case FALLING: {
+                wakeUpPinMode = INPUT_PULLUP;
+                break;
+            }
+            case CHANGE:
+            default:
+                wakeUpPinMode = INPUT;
+                break;
+        }
+
+        HAL_Pin_Mode(wakeUpPin, wakeUpPinMode);
+        HAL_InterruptExtraConfiguration irqConf = {0};
+        irqConf.version = HAL_INTERRUPT_EXTRA_CONFIGURATION_VERSION_2;
+        irqConf.IRQChannelPreemptionPriority = 0;
+        irqConf.IRQChannelSubPriority = 0;
+        irqConf.keepHandler = 1;
+        irqConf.keepPriority = 1;
+        HAL_Interrupts_Attach(wakeUpPin, wakeup_gpiote_handler, (void*)(uint32_t)i, edgeTriggerMode, &irqConf);
+
+        exit_conditions |= STOP_MODE_EXIT_CONDITION_PIN;
+    }
+
+    // Configure RTC wake-up
+    if (seconds > 0) {
+        wakeup_from_rtc(seconds);
+
+        exit_conditions |= STOP_MODE_EXIT_CONDITION_RTC;
+    }
+
+    // Enable all interrupt, __enable_irq()
+    err_code = sd_nvic_critical_region_exit(dummy);
+    APP_ERROR_CHECK(err_code);
+
+    fpu_sleep_prepare();
+
+    // Enter sleep mode, wait for event
+    do {
+        SPARK_ASSERT(sd_app_evt_wait() == NRF_SUCCESS);
+    } while (wakeup_info.source == STOP_MODE_EXIT_CONDITION_NONE);
+
+    int32_t reason = SYSTEM_ERROR_NOT_SUPPORTED;
+
+    if (exit_conditions & STOP_MODE_EXIT_CONDITION_PIN) {
+        if (wakeup_info.source & STOP_MODE_EXIT_CONDITION_PIN) {
+            reason = wakeup_info.pin_index + 1;
+        }
+        for (int i = 0; i < pins_count; i++) {
+            pin_t wakeUpPin = pins[i];
+            /* Detach the Interrupt pin */
+            HAL_Interrupts_Detach_Ext(wakeUpPin, 1, NULL);
+        }
+    }
+
+    if (exit_conditions & STOP_MODE_EXIT_CONDITION_RTC) {
+        if (wakeup_info.source & STOP_MODE_EXIT_CONDITION_RTC) {
+            reason = 0;
+        }
+    }
+
+    // Restore wakeup RTC
+    nrfx_rtc_uninit(&m_rtc);
+
+    // Restore GPIOTE
+    HAL_Interrupts_Restore();
+
+    // Restore hfclk
+    if (hfclk_resume) {
+        nrf_drv_clock_hfclk_request(NULL);
+        while (!nrf_drv_clock_hfclk_is_running()) {
+            ;
+        }
+    }
+
+    hal_exflash_init();
+
+    nrf5AlarmInit();
+
+    HAL_USB_Attach();
+
+    SysTick->CTRL |= SysTick_CTRL_ENABLE_Msk;
```

Comment:
We should block all interrupts until this point, otherwise a user interrupt handler will not be executed correctly if the wakeup pin was configured with `attachInterrupt()` in user firmware.